src/herder/TransactionQueue.cpp

```diff
@@ -30,183 +26,252 @@ TransactionQueue::TransactionQueue(Application& app, int pendingDepth,
-    for (auto const& tx : dropTxs)
+    // Note: start > 0 because every transaction in the TransactionQueue is
+    //       valid, and valid transactions have positive sequence numbers.
+    //       seqNum is unconstrained user input.
+    int64_t seqNum = tx->getSeqNum();
+    int64_t start = transactions.front().front()->getSeqNum();
+    if (seqNum >= start && (seqNum - start) < transactions.size())
     {
-        for (auto const& extracted : extract(tx, false).second)
+        auto iter = transactions.begin() + (seqNum - start);
+        if (iter->empty() || iter->front()->getSeqNum() != seqNum)
+        {
+            throw std::runtime_error("unexpected transaction queue state");
+        }
+        if (iter->front()->getInnerHash() == tx->getInnerHash())
         {
-            bannedFront.insert(extracted->getFullHash());
+            return iter;
         }
     }
+    return transactions.end();
 }
 
-bool
-TransactionQueue::contains(TransactionFrameBasePtr tx)
+int64_t
+TransactionQueue::getTotalFee(AccountID const& feeSource) const
 {
-    return find(tx).first != std::end(mPendingTransactions);
+    auto feeIter = mTotalFees.find(feeSource);
+    return (feeIter == mTotalFees.end()) ? 0 : feeIter->second;
 }
 
-TransactionQueue::FindResult
-TransactionQueue::find(TransactionFrameBasePtr const& tx)
+TransactionQueue::AddResult
+TransactionQueue::validateAndMaybeAddTransaction(
+    TransactionFrameBasePtr const& tx, int64_t validationSeqNum)
 {
-    auto const& acc = tx->getSourceID();
-    auto accIt = mPendingTransactions.find(acc);
-    if (accIt == std::end(mPendingTransactions))
+    LedgerTxn ltx(mApp.getLedgerTxnRoot());
+    if (!tx->checkValid(ltx, validationSeqNum))
     {
-        return {std::end(mPendingTransactions), {}};
+        return AddResult::ADD_STATUS_ERROR;
     }
 
-    auto& txs = accIt->second.mTransactions;
-    auto txIt =
-        std::find_if(std::begin(txs), std::end(txs), [&](auto const& t) {
-            return tx->getSeqNum() == t->getSeqNum();
-        });
-    if (txIt == std::end(txs))
+    int64_t totalFees = getTotalFee(tx->getFeeSourceID());
+    auto feeSource = stellar::loadAccount(ltx, tx->getFeeSourceID());
+    if (INT64_MAX - totalFees < tx->getFeeBid() ||
+        getAvailableBalance(ltx.loadHeader(), feeSource) <
+            totalFees + tx->getFeeBid())
     {
-        return {std::end(mPendingTransactions), {}};
+        tx->getResult().result.code(txINSUFFICIENT_BALANCE);
+        return AddResult::ADD_STATUS_ERROR;
     }
 
-    if ((*txIt)->getFullHash() != tx->getFullHash())
+    addFee(tx);
+
+    auto& transactions = mPendingTransactions[tx->getSourceID()].mTransactions;
+    auto transactionsIter = findTx(transactions, tx);
+    if (transactionsIter != transactions.end())
     {
-        return {std::end(mPendingTransactions), {}};
+        transactionsIter->emplace_back(tx);
+    }
+    else
+    {
+        transactions.emplace_back(1, tx);
```

Comment:
maybe consider using
`        transactions.emplace_back(std::vector<TransactionFrameBasePtr>{tx});`
(easier to read than `transactions.emplace_back(1, tx);`)